src/Storages/StorageMerge.cpp

```diff
@@ -1001,6 +1047,43 @@ void ReadFromMerge::convertingSourceStream(
     {
         return std::make_shared<ExpressionTransform>(stream_header, actions);
     });
+
+
+    auto row_policy_filter = local_context->getRowPolicyFilter(database_name, table_name, RowPolicyFilterType::SELECT_FILTER);
+
+    if (row_policy_filter)
+    {
+        ASTPtr expr = row_policy_filter->expression;
+
+        auto syntax_result = TreeRewriter(local_context).analyze(expr, pipe_columns);
+        auto expression_analyzer = ExpressionAnalyzer{row_policy_filter->expression, syntax_result, local_context};
+
+        auto actions_dag = expression_analyzer.getActionsDAG(true, false);
+        auto filter_actions = std::make_shared<ExpressionActions>(actions_dag, ExpressionActionsSettings::fromContext(local_context, CompileExpressions::yes));
+        auto required_columns = filter_actions->getRequiredColumns();
+
+        LOG_TRACE(&Poco::Logger::get("ReadFromMerge::convertinfSourceStream"), "filter_actions_dag: {},<> {}, <> {}",
+            filter_actions->getActionsDAG().dumpNames(), filter_actions->getActionsDAG().dumpDAG(), filter_actions->getSampleBlock().dumpStructure());
+
+
+        auto fa_actions_columns_sorted = filter_actions->getSampleBlock().getNames();
+        std::sort(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end());
+
+        Names required_columns_sorted = required_columns;
+        std::sort(required_columns_sorted.begin(), required_columns_sorted.end());
+
+        Names filter_columns;
+
+
+        std::set_difference(fa_actions_columns_sorted.begin(), fa_actions_columns_sorted.end(),
+            required_columns.begin(), required_columns.end(),
+            std::inserter(filter_columns, filter_columns.begin()));
```

Comment:
This looks almost identical to the code block you've added above, maybe factor it out into a function?